### DISCUSSION TITLE
Allow addition of custom oAuth providers.

### DISCUSSION THREAD

**Discussion by sandiprb** [strapi/strapi#6197] [Legacy Topics]
I'm using Straps for a front-end React application. This is one of the many applications handled by the enterprise.

The organization uses a FedAuth SSO login. We want the same to be used in this application too.

Upon login on the FedAuth, it redirects to my React application callback with a JWT token.

- Can I create a custom OAuth provider which can work for my custom FedAuth URLs.

*or*

- Can I create users with the email (extracted from JWT) in Strapi without a password?

**Comment by lauriejim** [comment]
Hello! According to this documentation, you will be able to create a custom provider that matchesFedAuth.

[documentation](https://strapi.io/documentation/3.0.0-beta.x/plugins/users-permissions.html#adding-a-new-provider-to-your-project)

**Reply by sandiprb** [reply]
Hey, @lauriejim So I  managed to achieve this with a custom controller. I created a custom controller which takes a jwt token returned by my fedauth login and then creates a user.  The custom controllers decrypt the token and check if a user exists in the system with given email, otherwise create a user and returns strapi token for the user.

**Reply by nitewulf** [reply]
@sandiprb I'm trying to do it sounds like exactly the same thing.  I've been fighting with resource/auth server integration for weeks.... If you wouldn't mind could you share the controller you made?

**Reply by sandiprb** [reply]
@nitewulf sure. Could you elaborate what exactly are you trying to achieve?

**Reply by sandiprb** [reply]
@nitewulf Also here's my  solutions

routes config as below
```
{
  "routes": [
    {
      "method": "POST",
      "path": "/ssoOAuth",
      "handler": "Auth.ssoOAuth"
    }
  ]
}
```

**Controller Path:** `extensions/users-permissions/controllers/Auth.js`

```
const jwt_decode = require("jwt-decode");

const formatError = (error) => [
  { messages: [{ id: error.id, message: error.message, field: error.field }] },
];

module.exports = {
  ssoOAuth: async (ctx) => {
    const advanced = await strapi
      .store({
        environment: "",
        type: "plugin",
        name: "users-permissions",
        key: "advanced",
      })
      .get();

    let userData;
    const { sso_token } = ctx.request.body; // notice this is a thirdparty service token
    let email;
    try {
      const jwtData = jwt_decode(sso_token);
      email = jwtData.email.toLowerCase();
    } catch (error) {
      // send a 400 response if unable to decode token
      ctx.badRequest(null, formatError(error));
    }

    // check if user exists with the same email
    userData = await strapi
      .query("user", "users-permissions")
      .findOne({ email });

    if (!userData) {
      console.debug("User not found creating a new user!");
      try {
        const { email, name, username } = jwt_decode(sso_token);

        const user = {
          email: email.toLowerCase(),
          username: username.toLowerCase(),
          name,
          confirmed: true,
          provider: "sso",
        };
        const defaultRole = await strapi
          .query("role", "users-permissions")
          .findOne({ type: advanced.default_role }, []);
        user.role = defaultRole.id;
        userData = await strapi.plugins["users-permissions"].services.user.add(
          user
        );
      } catch (error) {
        console.log(error);
        ctx.badRequest(null, formatError(error));
      }
    }
    return ctx.send({
      jwt: strapi.plugins["users-permissions"].services.jwt.issue({
        id: userData.id,
      }),
      user: strapi.admin.services.auth.sanitizeUser(userData),
    });
  },
};
```

**Reply by nitewulf** [reply]
The ultimate goal is basically my other post.  I'm trying to get Orchard Core to act as my Auth server and be able to serve up content from  Stapi to android clients (inventory warehouse users primarily).  The users currently use their Orchard login to access a Web portal full of a bunch of reports and other functions.  Strapi was going to be the host of Inventory data using the same login.  Combing the two has been difficult.  They (orchard and strapi) are so both developer-focused that I've come to believe you have to build 80% of what you want and .NET  is my background, unfortunately, I've been struggling with integrating them.

**Reply by nitewulf** [reply]
thank you for the controller.  I see the logic in which it functions.  I'm going to work on this week seeing if I can intercept the token from orchard and generate a strapi user with the permissions I want the user to have and send back a new JWT to my android client app using a custom controller based on your work.  Thanks.
